datafusion/expr/src/logical_plan/plan.rs

```diff
@@ -467,6 +468,200 @@ impl LogicalPlan {
         self.with_new_exprs(self.expressions(), inputs.to_vec())
     }
 
+    /// Recomputes schema and type information for this LogicalPlan if needed.
```

Comment:
I believe this is a new API for using `TreeNode` to rewrite plans in ways that change the schema. 

This effectively factors out the recalculation part of `LogicalPlan::new_with_exprs`

Reply:
FYI @peter-toth  I suspect you may need something like this for common subexpression elimination / https://github.com/apache/datafusion/issues/9873